### PR TITLE
pkg/cpio, cmds/core/cpio: Allow reading from pipes

### PR DESCRIPTION
**Pull request by rminnich**
Currently, the cpio package can not read from pipes. This violates
a fundamental property of both cpio and Unix in general.

The fix might seem to be allowing creation of Readers from an os.File.
We could then see if the File in question implements Reader or
ReaderAt: if ok, rat := f.(ReaderAt) ...
and handle each case, transparent to the caller.

The problem is that some Go os.File types, such as Pipes,
advertise themselves as implementing ReaderAt but in fact do not
do so.

e.g.
	r, _, _ := os.Pipe()
	var b [2]byte
	var _ io.ReaderAt = r
	_, err := r.ReadAt(b[:], 0)
	fmt.Printf("%v", err)

will fail, since whatever Go may think, Linux pipes do not allow a
seek, even when the offset is not changing. So the test
above will not work.

The Go runtime needs to be fixed, so that it does not claim
pipes implement io.ReaderAt, but it's going to be a long
time before such a change propagates everywhere. The cpio package
must accomodate this error. Further, there will almost
certainly be other future types which make this mistake.

Provide a NewFileReader function, which accepts *os.File
and returns a Reader. The function performs a simple test,
io.Seek(0,0), and, if that works, just creates a Reader
using the File's ReadAt.

If that seek fails, the function wraps the file in a
discardreader.

It is best to use a ReadAt when possible to avoid having
to read and discard data to implement seek.

cpio can now use stdin from a pipe.

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

### CODE REVIEW

**Comment by rminnich** [comment]
we need this for Harvey, among other things.

File: pkg/cpio/newc.go [hunk 1]
```diff
@@ -183,19 +185,82 @@ type reader struct {
 	pos int64
 }
 
+type discarder struct {
```
Comment (gmacd) [inline pkg/cpio/newc.go:188]: It wasn't obvious to me that a discarder exists to provide an implementation of ReadAt where Seek isn't supported.  Might be worth a comment.

**Review by gmacd — DISMISSED** [review]

File: pkg/cpio/newc.go (hunk 1 above)
Comment (rminnich) [reply pkg/cpio/newc.go:192]: done.

File: pkg/cpio/newc.go [hunk 2]
```diff
@@ -183,19 +185,86 @@ type reader struct {
...
+	r   io.Reader
+	pos int64
+}
+
+// ReadAt implements ReadAt for a discarder.
+// It is an error for the offset to be negative.
+func (r *discarder) ReadAt(p []byte, off int64) (int, error) {
+	if off-r.pos < 0 {
+		return -1, fmt.Errorf("negative seek on discarder not allowed")
```
Comment (fhs) [inline pkg/cpio/newc.go:201]: I think the convention is to return 0 for error.
```
$ go doc io ReaderAt
...
It returns the number of bytes read (0 <= n <= len(p)) and any
error encountered.
...
````

```suggestion
		return 0, fmt.Errorf("negative seek on discarder not allowed")
```

File: pkg/cpio/newc.go [hunk 3]
```diff
@@ -183,19 +185,86 @@ type reader struct {
...
+	}
+	if off != r.pos {
+		i, err := io.Copy(ioutil.Discard, io.LimitReader(r.r, off-r.pos))
+		if err != nil || i != off-r.pos {
+			return -1, err
```
Comment (fhs) [inline pkg/cpio/newc.go:206]: ditto
```suggestion
			return 0, err
```

**Review by gmacd — APPROVED** [review]

**Pull request merged**